platforms/jvm/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java

```diff
@@ -263,29 +262,20 @@ private static void configureCompileDefaults(final Project project, final ScalaR
             compile.getScalaClasspath().convention(project.provider(() -> scalaRuntime.inferScalaClasspath(compile.getClasspath())));
             compile.getZincClasspath().convention(project.getConfigurations().getAt(ZINC_CONFIGURATION_NAME));
             compile.getScalaCompilerPlugins().convention(project.getConfigurations().getAt(SCALA_COMPILER_PLUGINS_CONFIGURATION_NAME));
-            ConventionMapping conventionMapping = compile.getConventionMapping();
-            conventionMapping.map("sourceCompatibility", () -> computeJavaSourceCompatibilityConvention(javaExtension, compile).toString());
-            conventionMapping.map("targetCompatibility", () -> computeJavaTargetCompatibilityConvention(javaExtension, compile).toString());
+            compile.getSourceCompatibility().convention(
+                javaExtension.getSourceCompatibility().map(JavaVersion::toString).orElse(
+                    compile.getJavaLauncher().map(launcher -> launcher.getMetadata().getLanguageVersion().toString())
+                )
+            );
+            compile.getTargetCompatibility().convention(
+                javaExtension.getTargetCompatibility().map(JavaVersion::toString).orElse(
+                    compile.getSourceCompatibility()
+                )
+            );
```

Comment:
I know this was prior behavior, but I don't think we should carry this along through the provider API migration. 

IMO we should handle the java launcher integration internally within `AbstractCompile` (or ScalaCompile/JavaCompile/GroovyCompile). It seems weird to have a convention of one property of a task to be linked to the value of another property of the task. 

I think it should be something like
```
compile.getSourceCompatibility().convention(javaExtension.getSourceCompatibility().map(JavaVersion::toString))
compile.getTargetCompatibility().convention(javaExtension.getTargetCompatibility().map(JavaVersion::toString))
```

and then _in the task action_ of the compile task, if `getSourceCompatibiltiy` is not set, then fall back to the java launcher's source/target compatibility. 

Similar to my prior comment, this is somewhat of a breaking change -- but if we're doing it in one place I think we should bite the bullet and do it here too, as it is better to not have properties whose values depend on downloading toolchains